File: solutions/0005-longest-palindromic-substring/solution.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        vector<vector<int>> dp(s.size(),vector<int>(s.size(),0));
        string maxi = "";
        for(int diag=0;diag<s.size();diag++){
            int starti = 0;
            int startj = diag;
            for(int step=0;step<s.size()-diag;step++){
                int  i = starti+step;
                int  j = startj+step;
                if(i==j){
                    dp[i][j]=1;
                }
                else if(i==j-1 and s[i]==s[j]){
                    dp[i][j] =1 ;
                }
                else if(s[i]==s[j]){
                    dp[i][j]=dp[i+1][j-1];
                }
                else{
                    dp[i][j]=0;
                }

                if(dp[i][j]==1 and maxi.size()<j-i+1){
                    maxi = s.substr(i,j-i+1);
                }
            }
        }
        return maxi;
    }
};

File: solutions/0005-longest-palindromic-substring/solution.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        dp = [[0]*len(s) for _ in s]
        for i in range(len(s)-1,-1,-1):
            for j in range(0,i+1):
                I = j
                J = len(s)-i-1+j
                if(I==J):
                    dp[I][J]= 1
                elif(I==J-1):
                    dp[I][J]= 1 if s[I]==s[J] else 0
                else:
                    dp[I][J] = 1 if (s[I]==s[J] and dp[I+1][J-1]==1) else 0
        # print(dp)
        maxLen = 0
        maxString = ""
        for i in range(0,len(s)):
            for j in range(0,len(s)):
                if dp[i][j]==1 and maxLen<=j-i:
                    maxLen = j-i
                    maxString = s[i:j+1]
                
        return maxString



            

            
        
